manager/controlapi/secret.go

```diff
@@ -12,13 +18,41 @@ import (
 // MaxSecretSize is the maximum byte length of the `Secret.Spec.Data` field.
 const MaxSecretSize = 500 * 1024 // 500KB
 
+var isValidSecretName = regexp.MustCompile(`^[a-zA-Z0-9](?:[a-zA-Z0-9-_.]{0,62}[a-zA-Z0-9])*$`)
+
+// assumes spec is not nil
+func secretFromSecretSpec(spec *api.SecretSpec) *api.Secret {
+	checksumBytes := sha256.Sum256(spec.Data)
+	return &api.Secret{
+		ID:         identity.NewID(),
+		Spec:       *spec,
+		SecretSize: uint32(len(spec.Data)),
+		Digest:     hex.EncodeToString(checksumBytes[:]),
```

Comment:
I suggest using the `github.com/docker/distribution/digest` package for this. `digest.FromBytes(spec.Data)` will give a result like `sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`. Including the digest algorithm as a prefix is nice because then we aren't implicitly tied to a particular format going forward. We use this package in a few other places in swarmkit.